R/geom-polygon.r

```diff
@@ -83,30 +102,60 @@ GeomPolygon <- ggproto("GeomPolygon", Geom,
     if (n == 1) return(zeroGrob())
 
     munched <- coord_munch(coord, data, panel_params)
-    # Sort by group to make sure that colors, fill, etc. come in same order
-    munched <- munched[order(munched$group), ]
 
-    # For gpar(), there is one entry per polygon (not one entry per point).
-    # We'll pull the first value from each group, and assume all these values
-    # are the same within each group.
-    first_idx <- !duplicated(munched$group)
-    first_rows <- munched[first_idx, ]
+    if (is.null(munched$subgroup)) {
+      # Sort by group to make sure that colors, fill, etc. come in same order
+      munched <- munched[order(munched$group), ]
+
+      # For gpar(), there is one entry per polygon (not one entry per point).
+      # We'll pull the first value from each group, and assume all these values
```

Comment:
Other geoms check this assumption and issue a warning if an aesthetic changes where it’s not supposed to. See e.g. `geom_area()`. Needed here?

Reply:
I think that may become prohibitly expensive to check for the kind of data that people will throw at this... Maybe make it more clear in the docs?

Reply:
Do you think a single call to `unique()` is that much more expensive than, say, the ordering of the data that is also happening?
https://github.com/tidyverse/ggplot2/blob/18be30e1515b2b9d05ade4f3f2b5051aa1336fcf/R/geom-ribbon.r#L80-L83

At a minimum, this might be worth a careful benchmark on a very large dataset, such as the output from `isobands()` on a large raster image.

Reply:
I’ll happily do the benchmark, but the big difference is that geom_ribbon uses a draw_group method whereas geom_polygon uses draw_panel. This means that geom_polygon would have to split up the data and call unique on each sub-data.frame all for the sake of a possible warning. For geom ribbon the split has already been done so it is rather cheap

Reply:
I see. In any case, it's your call. I won't insist.

Reply:
We may wanted to eventually tackle this problem by providing a new geom where there's one row per polygon and the vertices are stored in a nested data frame.